pkg/reaper/deployment.go

```diff
@@ -412,3 +519,92 @@ func getPodMeta(reaper *api.Reaper) meta.Tags {
 		Annotations: podAnnotations,
 	}
 }
+
+func MakeActualDeploymentType(actualReaper *api.Reaper) (client.Object, error) {
+	if actualReaper.Spec.StorageType == api.StorageTypeCassandra {
+		return &appsv1.Deployment{}, nil
+	} else if actualReaper.Spec.StorageType == api.StorageTypeMemory {
+		return &appsv1.StatefulSet{}, nil
+	} else {
+		err := fmt.Errorf("invalid storage type %s", actualReaper.Spec.StorageType)
+		return nil, err
+	}
+}
+
+func MakeDesiredDeploymentType(actualReaper *api.Reaper, dc *cassdcapi.CassandraDatacenter, keystorePassword *string, truststorePassword *string, logger logr.Logger, authVars ...*corev1.EnvVar) (client.Object, error) {
+	if actualReaper.Spec.StorageType == api.StorageTypeCassandra {
+		return NewDeployment(actualReaper, dc, keystorePassword, truststorePassword, logger, authVars...), nil
+	} else if actualReaper.Spec.StorageType == api.StorageTypeMemory {
+		// we're checking for this same thing in the k8ssandra-cluster webohooks
+		// but in tests (and in the future) we'll be creating reaper directly (not through k8ssandra-cluster)
+		// so we need to double-check
+		if actualReaper.Spec.StorageConfig == nil {
+			err := fmt.Errorf("storageConfig is required for memory storage")
+			return nil, err
+		}
+		return NewStatefulSet(actualReaper, dc, logger, keystorePassword, truststorePassword, authVars...), nil
+	} else {
+		err := fmt.Errorf("invalid storage type %s", actualReaper.Spec.StorageType)
+		return nil, err
+	}
+}
+
+func DeepCopyActualDeployment(actualDeployment client.Object) (client.Object, error) {
+	switch actual := actualDeployment.(type) {
+	case *appsv1.Deployment:
+		actualDeployment = actual.DeepCopy()
+		return actualDeployment, nil
+	case *appsv1.StatefulSet:
+		actualDeployment = actual.DeepCopy()
+		return actualDeployment, nil
+	default:
+		err := fmt.Errorf("unexpected type %T", actualDeployment)
+		return nil, err
+	}
+}
+
+func EnsureSingleReplica(actualReaper *api.Reaper, actualDeployment client.Object, desiredDeployment client.Object, logger logr.Logger) error {
+	if actualReaper.Spec.StorageType == api.StorageTypeMemory {
+		desiredReplicas := getDeploymentReplicas(desiredDeployment, logger)
+		if desiredReplicas > 1 {
+			logger.Info(fmt.Sprintf("reaper with memory storage can only have one replica, not allowing the desired %d", desiredReplicas))
+			if err := setDeploymentReplicas(&desiredDeployment, &api.SingleReplica); err != nil {
+				return err
+			}
+		}
+		actualReplicas := getDeploymentReplicas(actualDeployment, logger)
+		if actualReplicas > 1 {
+			logger.Info(fmt.Sprintf("reaper with memory storage currently has %d replicas, scaling down to 1", actualReplicas))
+			if err := setDeploymentReplicas(&desiredDeployment, &api.SingleReplica); err != nil {
```

Comment:
nitpick: I don't understand the &api.SingleReplica usage.. if it's 1, then why not simply do ``ptr.To[int32](1)`` ?

Reply:
Another no-good reason. Fixed.